### PR TITLE
Implement platform-specific options for rdctl set|start

### LINKED ISSUE #3806: Incorrect auto-generated options in MacOS
### Actual Behavior

By testing the auto-generated options for `rdctl set`, the following flags must not be available for rdctl set: `--kubernetes.host-resolver` as this option applies just for Windows and `--container-engine.image-allow-list.locked` as it cannot be changed via API.

### Steps to Reproduce

1.Run `rdctl set --kubernetes.host-resolver`
2. Check the output
3. Run `rdctl set --container-engine.image-allow-list.locked`
4. Check the output

### Result

```
Error: errors in attempt to update settings:
Changing field containerEngine.imageAllowList.locked via the API isn't supported
```

### Expected Behavior

the following flags must not be available for rdctl set: `--kubernetes.host-resolver` as this option applies just for Windows and `--container-engine.image-allow-list.locked` as it cannot be changed via API.

### Additional Information

_No response_

### Rancher Desktop Version

1.7.0-133-g25e1eabd

### Rancher Desktop K8s Version

1.17.17

### Which container engine are you using?

containerd (nerdctl)

### What operating system are you using?

macOS

### Operating System / Build Version

macOS Monterey

### What CPU architecture are you using?

arm64 (Apple Silicon)

### Linux only: what package format did you use to install Rancher Desktop?

None

### Windows User Only

_No response_

### PR DESCRIPTION
**Pull request by ericpromislow**
Fixes #3806

### CODE REVIEW

**Comment by jandubois** [comment]
I think it would be better if the options would still be implemented, but hidden, and when used would show an error like "The --foo.bar option is not supported on the BeOS platform" instead of showing an "Unknown option" error.

**Comment by ericpromislow** [comment]
I'll have to experiment to see if hidden options can work that way (with a semi-custom message)

**Comment by jandubois** [comment]
> I'll have to experiment to see if hidden options can work that way (with a semi-custom message)

You just generate an additional condition in `UpdateFieldsForJSON`, instead of

```golang
        if flags.Changed("kubernetes.memory-in-gb") {
                specifiedSettingsForJSON.Kubernetes.MemoryInGB = &specifiedSettings.Kubernetes.MemoryInGB
                changedSomething = true
        }
```

you will generate this (on Windows):

```golang
        if flags.Changed("kubernetes.memory-in-gb") {
                return nil, errors.New("--kubernetes.memory-in-gb is not supported on Windows")
        }
```

**Comment by ericpromislow** [comment]
Can I get agreement that the platform names in the error messages should be "Windows", "Linux", and "macOS"?

Also I see I need a stealth ticket to port RD to BeOS :)

**Comment by jandubois** [comment]
> Can I get agreement that the platform names in the error messages should be "Windows", "Linux", and "macOS"?

SGTM

> Also I see I need a stealth ticket to port RD to BeOS :)

It doesn't already work there? 😄

**Comment by jandubois** [comment]
Now, I know that the Linux version of `rdctl` doesn't properly work under WSL anyways, but do we ship a custom build for WSL there, or is it a copy of the regular Linux one?

If we already do build a custom version, then this one needs to support the Windows-specific options, and not the Linux ones.

**Comment by ericpromislow** [comment]
ok, this code is wrong -- the platform test needs to be done at runtime, not code-generation time

**Comment by ericpromislow** [comment]
Note that we can generate the options at code-gen-time, because `rdctl set` has the same constraints on a windows box whether you run `rdctl.exe` or `wsl -d ... rdctl`, but we still need to do the hidden-option thing

File: e2e/rdctl.e2e.spec.ts [hunk 1]
```diff
@@ -695,6 +695,24 @@ test.describe('Command server', () => {
     });
 
     test.describe('set', () => {
+      const unsupportedPrefsByPlatform = {
```
Comment (mook-as) [inline e2e/rdctl.e2e.spec.ts:698]: Consider typing this more explicitly so it's easier to reason about:
```suggestion
      const unsupportedPrefsByPlatform: {[x in NodeJS.Platform]?: [string, any][]} = {
```

This means line 714 can become:
```ts
      const unsupportedOptions = unsupportedPrefsByPlatform[process.platform] ?? [];
```

File: e2e/rdctl.e2e.spec.ts [hunk 2]
```diff
@@ -855,6 +873,16 @@ test.describe('Command server', () => {
           expect(result.stderr).toEqual('');
         });
       });
+
+      test('complains about options not intended for current platform', async() => {
+        // See https://github.com/microsoft/playwright/issues/7036 for the explanation
+        // why I'm not doing `test.describe(... set up the array ... test.each(array)...
```
Comment (mook-as) [inline e2e/rdctl.e2e.spec.ts:879]: So, in summary: playwright doesn't support `test.each`? That seems like a more concise and clear comment…

File: pkg/rancher-desktop/assets/specs/command-api.yaml [hunk 3]
```diff
@@ -238,11 +238,13 @@ components:
           properties:
             adminAccess:
               type: boolean
+              x-rd-platforms: ['darwin', 'linux'] # Only in the specified platforms
```
Comment (mook-as) [inline pkg/rancher-desktop/assets/specs/command-api.yaml:241]: I'm starting to wonder if we can have a schema for the schema :p (No need to worry about it in this PR, though.)

File: scripts/generateCliCode.ts [hunk 4]
```diff
@@ -258,16 +265,23 @@ class Generator {
...
 
-  protected walkProperty(propertyName: string, preference: yamlObject, settingsTree: settingsTreeType): void {
+  protected walkProperty(
+    propertyName: string,
+    preference: yamlObject,
+    notAvailable: boolean,
+    settingsTree: settingsTreeType): void {
+    const platforms = preference['x-rd-platforms'] ?? [];
+
+    notAvailable ||= platforms.length > 0 && !platforms.includes(os.platform());
```
Comment (mook-as) [inline scripts/generateCliCode.ts:275]: FWIW, `os.platform()` is also available as `process.platform` (a property off a global variable), which means you don't need to import it.

File: scripts/generateCliCode.ts [hunk 5]
```diff
@@ -258,16 +265,23 @@ class Generator {
     }
   }
 
-  protected walkProperty(propertyName: string, preference: yamlObject, settingsTree: settingsTreeType): void {
+  protected walkProperty(
+    propertyName: string,
+    preference: yamlObject,
+    notAvailable: boolean,
+    settingsTree: settingsTreeType): void {
```
Comment (mook-as) [inline scripts/generateCliCode.ts:272]: There's enough parameters that I'm starting to wonder if we should pass in an object :) (Don't need to do anything yet, just something to consider for later.)

**Review by mook-as — CHANGES_REQUESTED** [review]
Mostly fine, just a comment change plus some suggestions that can be turned down.

File: e2e/rdctl.e2e.spec.ts (hunk 1 above)
Comment (ericpromislow) [reply e2e/rdctl.e2e.spec.ts:691]: ok

File: pkg/rancher-desktop/assets/specs/command-api.yaml (hunk 3 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/assets/specs/command-api.yaml:241]: ok

File: scripts/generateCliCode.ts (hunk 5 above)
Comment (ericpromislow) [reply scripts/generateCliCode.ts:272]: It's only a problem with `updateLeaf`, and I would just put the variants of property names, types, and values in an object, and leave preference`, `notAvailable`, and `settingsTree` as separate parameters.

Comment (ericpromislow) [reply scripts/generateCliCode.ts:271]: I tested it out. The change adds 44 lines and removes 13.

The three calls to `updateLeaf` call a function to wrap five values in an object.

`updateLeaf` then peels them out of that object.

It doesn't seem worth it to me.

Comment (mook-as) [reply scripts/generateCliCode.ts:271]: Seems reasonable. I was hoping it would make for a better API, but if not then it isn't.

**Review by mook-as — DISMISSED** [review]

**Review by mook-as — APPROVED** [review]

**Pull request merged**